apps/www/content/docs/dark-mode/tanstack-start.mdx

```diff
@@ -0,0 +1,241 @@
+---
+title: Tanstack Start
+description: Adding dark mode to your tanstack start app.
+---
+
+## Dark mode
+
+<Steps>
+
+### Create a theme provider
+
+```tsx title="components/theme-provider.tsx"
+import { createServerFn } from '@tanstack/react-start';
+import { getCookie, setCookie } from '@tanstack/react-start/server';
+import {
+  createContext,
+  useContext,
+  useEffect,
+  useMemo,
+  useState
+} from 'react';
+
+type Theme = 'dark' | 'light' | 'system';
+
+type ThemeProviderProps = {
+  children: React.ReactNode;
+  defaultTheme?: Theme;
+  storageKey?: string;
+};
+
+type ThemeProviderState = {
+  theme: Theme;
+  toggleTheme: (theme: Theme) => void;
+};
+
+const initialState: ThemeProviderState = {
+  theme: 'system',
+  toggleTheme: () => null,
+};
+
+export const getThemeCookie = createServerFn({ method: 'GET' })
+  .validator((d: { storageKey: string; defaultTheme?: Theme }) => d)
+  .handler(async ({ data }) => {
+    const { storageKey, defaultTheme = 'light' } = data;
+    const cookieTheme = getCookie(data.storageKey) as Theme | undefined;
+
+    return cookieTheme ?? defaultTheme;
+  });
+
+export const updateThemeCookie = createServerFn({ method: 'POST' })
+  .validator((d: { theme: Theme; storageKey: string }) => d)
+  .handler(async ({ data }) => {
+    const { storageKey, theme } = data;
+
+    setCookie(storageKey, theme, {
+      httpOnly: false,
+      sameSite: 'lax',
+      secure: process.env.NODE_ENV === 'production',
+      path: '/',
+      maxAge: 60 * 60 * 24 * 365 * 10,
+    });
+  });
+
+const ThemeContext = createContext<ThemeProviderState>(initialState);
+
+export function ThemeProvider({
+  children,
+  defaultTheme = 'system',
+  storageKey = 'tanstack-start-ui-theme',
+}: ThemeProviderProps) {
+  const [theme, setTheme] = useState(defaultTheme);
+
+  useEffect(() => {
+    const root = window.document.documentElement;
+
+    root.classList.remove('light', 'dark');
+
+    if (typeof window !== 'undefined' && theme === 'system') {
+      const systemTheme = window.matchMedia('(prefers-color-scheme: dark)')
+        .matches
+        ? 'dark'
+        : 'light';
+
+      root.classList.add(systemTheme);
+    } else {
+      root.classList.add(theme);
+    }
+
+    updateThemeCookie({ data: { storageKey, theme } });
+  }, [theme, storageKey]);
+
+  const value = useMemo<ThemeProviderState>(
+    () => ({
+      theme,
+      toggleTheme: setTheme,
+    }),
+    [theme]
+  );
+
+  return (
+    <ThemeContext.Provider value={value}>{children}</ThemeContext.Provider>
+  );
+}
+
+export function useTheme() {
+  const ctx = useContext(ThemeContext);
+
+  if (!ctx) {
+    throw new Error('useTheme must be used within ThemeProvider');
+  }
+
+  return ctx;
+}
+```
+
+### Get the theme cookie and wrap your app with ThemeProvider
+
+Get the theme cookie in the loader.
+
+```tsx title="routes/__root.tsx"
```

Comment:
It could be a lot more useful to show a diff view here instead as I missed a single line to change when comparing this to my file. An example of it being used in the docs already is here: https://ui.shadcn.com/docs/components/pagination#nextjs
https://github.com/shadcn-ui/ui/blob/04432835f91dbe0a3273f50ac829e0a377110e5f/apps/v4/content/docs/components/pagination.mdx?plain=1#L83-L99